user/fixsccs.man
.\" $Id: fixsccs.man,v 6.1 2023/01/20 23:18:28 tom Exp $
.de Es
.ne \\$1
.nr mE \\n(.f
.RS 5n
.sp .7
.nf
.nh
.ta 9n 17n 25n 33n 41n 49n
.ft CW
..
.de Eh
.ft \\n(mE
.fi
.hy \\n(HY
.RE
.sp .7
..
.TH FIXSCCS 1
.SH NAME
.PP
\fBfixsccs\fR \- SCCS repair utility
.SH USAGE
.PP
\fBfixsccs\fI [options] [file-specifications]\fR
.SH SYNOPSIS
Sets SCCS archive file times to match the checkin time of the tip version.
.PP
.SH DESCRIPTION
Touchs sccs files so that their modification times match the
checkin-time recorded for the tip version.
This is used to accommodate SunOS's make facility,
which has a get-from-sccs rule.
Normally, \fBputdelta\fR does this;
\fBfixsccs\fR is used to
handle the cases when someone bypasses \fBputdelta\fR and uses
\fBadmin\fR or \fBdelta\fR directly.
.SH OPTIONS
.TP
.B \-a
process all directories, including those beginning with \".\"
.TP
.B \-n
scan archives only, do not try to set modification-times
.TP
.B \-q
quiet
.SH ENVIRONMENT
.PP
\fBFixsccs\fR is a standalone utility, which does not directly depend upon
the \fIsccs\fP applications.
.SH FILES
.PP
\fBFixsccs\fR consists of a single binary file written in C.
It runs in a POSIX environment.
.SH ANTICIPATED CHANGES
.PP
None.
.SH SEE ALSO
.PP
getdelta, putdelta, sccsget, sccsput, get\ (1)
.SH AUTHOR
.PP
Thomas E. Dickey


CHANGES
2023-01-21  Thomas E. Dickey  <dickey@invisible-island.net>

	* bin/makefile.in:
	s/USER/AS_USER/ to avoid environment-variable conflict

	* package/sccs_tools-12.0.spec: amended the version-dependency

2023-01-20  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/getdelta/test/run_test.sh, src/sccs2rcs/test/run_test.sh,
	  src/sccs2rcs/test/do_test.sh:
	shellcheck warnings

	* src/putdelta/test/run_test.sh: fix shellcheck warnings

	* src/putdelta/test/run_test.sh:
	discard stderr (some variants complain profusely about missing keywords)

	* src/getdelta/test/run_test.sh:
	modify to ensure the files have SCCS keywords, since some variants do not handle this gracefully

	* user/sccs2rcs.man: check-manpage

	* user/fixsccs.man: RCS_BASE

	* user/sccsget.man, user/sccsput.man, user/getdelta.man,
	  user/putdelta.man:
	check-manpage

	* user/makefile.in: add manpage for fixsccs

	* package/debian/changelog, package/sccs_tools-12.0.spec: bump

	* package/debian/control: update standards

	* package/debian/compat: bump

2023-01-19  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/sccs_tools-12.0.spec:
	use td_lib package, rather than side-by-side configuration

	* configure: regen

	* aclocal.m4: provide for finding td_lib using pkg-config

	* package/debian/copyright, COPYING: bump

2022-12-31  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/changelog, package/sccs_tools-12.0.spec: bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

2022-10-11  Thomas E. Dickey  <dickey@invisible-island.net>

	* user/makefile.in: use install-sh if needed

	* src/putdelta/test/run_test.sh: s/fgrep/grep/

	* package/debian/copyright: bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

	* COPYING, package/debian/changelog, package/sccs_tools-12.0.spec: bump

2022-08-01  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess, config.sub: 2022-08-01

2021-03-28  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/changelog, package/sccs_tools-12.0.spec: bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

2021-01-10  Thomas E. Dickey  <dickey@invisible-island.net>

	* configure: regen

	* aclocal.m4: quoting fixes for shellcheck

	* src/getdelta/src/getdelta.c, src/getdelta/src/sccsget.c,
	  src/sccs2rcs/src/sccs2rcs.c:
	gcc-warnings

	* configure: regen

	* aclocal.m4: resync with my-autoconf

	* COPYING, package/debian/changelog, package/debian/copyright,
	  package/sccs_tools-12.0.spec:
	bump

2020-12-08  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/changelog, package/sccs_tools-12.0.spec: bump

2020-12-07  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/getdelta/test/makefile.in, src/putdelta/test/makefile.in,
	  src/sccs2rcs/test/makefile.in, src/fixsccs/test/makefile.in,
	  src/getdelta/src/makefile.in, src/putdelta/src/makefile.in,
	  src/sccs2rcs/src/makefile.in, src/fixsccs/src/makefile.in:
	use VPATH

2020-12-06  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/copyright, COPYING, package/debian/changelog,
	  package/sccs_tools-12.0.spec:
	bump

2020-11-30  Thomas E. Dickey  <dickey@invisible-island.net>

	* install-sh: update to 2020-11-14

2020-11-26  Thomas E. Dickey  <dickey@invisible-island.net>

	* configure: regen

	* aclocal.m4: resync with my-autoconf

	* configure.in: without-X

2020-08-16  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess, config.sub: 2020-08-17

2019-12-21  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/putdelta/src/sccsput.c: suppress pager if not a tty

	* package/debian/changelog, package/sccs_tools-12.0.spec: bump

	* src/putdelta/src/putdelta.c: add an fflush

2019-12-06  Thomas E. Dickey  <dickey@invisible-island.net>

	* user/putdelta.man: mention the other options that putdelta handles

	* src/sccs2rcs/src/sccs2rcs.c: stricter gcc warnings

	* src/putdelta/src/putdelta.c, src/putdelta/src/sccsput.c:
	use DYN-argv lists to replace catarg()

	* package/debian/compat: bump

	* src/getdelta/src/getdelta.c: use DYN-argv lists to replace catarg()

	* user/sccsget.man, user/sccsput.man, user/putdelta.man,
	  user/getdelta.man:
	check-manpage fixes (rename Ex to Eh for consistency, other errata)

	* src/putdelta/test/run_test.sh: use sccs wrapper

	* src/getdelta/src/sccsget.c: use DYN-argv lists to replace catarg()

	* src/getdelta/test/run_test.sh: use sccs wrapper

	* configure: regen

	* aclocal.m4: resync with my-autoconf

	* package/debian/copyright, COPYING, package/debian/changelog,
	  package/sccs_tools-12.0.spec:
	bump

2019-09-11  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess, config.sub: 2019-09-10

2018-03-24  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/changelog: bump

	* package/sccs_tools-12.0.spec:
	bump, disable debug-packages, update ftp url

	* package/debian/watch: update ftp url

	* package/debian/rules: use Debian build-flags

2018-01-07  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/copyright, COPYING, package/debian/changelog,
	  package/sccs_tools-12.0.spec:
	bump

	* configure: regen

	* aclocal.m4, configure.in: resync with my-autoconf

2017-11-25  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: 2017-11-07

	* config.sub: 2017-11-23

2015-07-04  Thomas E. Dickey  <dickey@invisible-island.net>

	* user/makefile.in: improve clean-rule

	* package/sccs_tools-12.0.spec, package/debian/copyright,
	  package/debian/changelog, COPYING:
	bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

2015-05-02  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: 2015-03-04

	* config.sub: 2015-03-08

2014-12-14  Thomas E. Dickey  <dickey@invisible-island.net>

	* user/sccsput.man, user/putdelta.man, user/getdelta.man,
	  user/sccsget.man:
	adapt/reuse macros from cm_tools to improve groff's translation to html

2013-12-07  Thomas E. Dickey  <dickey@invisible-island.net>


src/fixsccs/src/fixsccs.c
/*
 * Title:	fixsccs.c (sccs put-tree)
 * Author:	T.E.Dickey
 * Created:	07 Jun 1994
 * Modified:
 *		
 * Function:	
 *		Touchs sccs files so that their modification times match the
 *		checkin-time recorded for the tip version.  This is used to
 *		accommodate SunOS's make facility, which has a get-from-sccs
 * 		rule. Normally, 'putdelta' does this; 'fixsccs' is used to
 *		handle the cases when someone bypasses 'putdelta' and uses
 *		'admin' or 'delta' directly.
 *
 * Options:	see 'usage()'
 */

#define	STR_PTYPES
#include	<ptypes.h>
#include	<sccsdefs.h>
#include	<time.h>	/* declares 'ctime()' */

MODULE_ID("$Id: fixsccs.c,v 6.13 2010/07/05 21:46:38 tom Exp $")

static int a_opt;		/* all-directory scan */
static int no_op;		/* no-op mode */
static int quiet;
static int fixed;

static void
FixSCCS(const char *path, const char *archive, Stat_t * sp)
{
    time_t date;
    const char *vers;
    const char *locker;

    sccslast(path, sccs2name(archive, TRUE), &vers, &date, &locker);
    if (*vers != '?'
	&& sp->st_mtime != date) {
	if (!quiet) {
	    PRINTF("%s %s\n", archive, vers);
	    PRINTF("   now: %s", ctime(&(sp->st_mtime)));
	    PRINTF("   set: %s", ctime(&date));
	}
	if (!no_op) {
	    if (setmtime(archive, date, (time_t) 0) < 0)
		failed(archive);
	}
	fixed++;
    }
}

/*ARGSUSED*/
static int
WALK_FUNC(scan_tree)
{
    static char prefix[] = SCCS_PREFIX;
    char tmp[MAXPATHLEN];
    char *leaf;

    (void) level;

    leaf = pathleaf(pathcat(tmp, path, name));

    if (sp == 0 || readable < 0) {
	readable = -1;
	perror(name);
    } else if (isDIR(sp->st_mode)) {
	if (!a_opt && !dotname(leaf) && *leaf == '.')
	    readable = -1;
	else if (!quiet)
	    track_wd(path);
    } else if (!strncmp(leaf, prefix, sizeof(prefix) - sizeof(char))
	       && isFILE(sp->st_mode)) {
	if (!quiet)
	    track_wd(path);
	FixSCCS(path, tmp, sp);
    } else
	readable = -1;

    return (readable);
}

static void
do_arg(const char *name)
{
    (void) walktree((char *) 0, name, scan_tree, "r", 0);
}

static
void
usage(void)
{
    static const char *tbl[] =
    {
	"Usage: fixsccs [options] files_or_directories"
	,""
	,"Sets SCCS archive file times to match the checkin time of the tip version."
	,""
	,"Options"
	,"  -a       process all directories, including those beginning with \".\""
	,"  -n       scan archives only, don't try to set mod-times"
	,"  -q       quiet"
	,""
    };
    unsigned j;
    for (j = 0; j < sizeof(tbl) / sizeof(tbl[0]); j++)
	FPRINTF(stderr, "%s\n", tbl[j]);
    exit(FAIL);
}

/*ARGSUSED*/
_MAIN
{
    register int j;

    while ((j = getopt(argc, argv, "anq")) != EOF) {
	switch (j) {
	case 'a':
	    a_opt = TRUE;
	    break;
	case 'n':
	    no_op = TRUE;
	    break;
	case 'q':
	    quiet = TRUE;
	    break;
	default:
	    usage();
	}
    }

    if (!quiet)
	track_wd((char *) 0);
    if (optind < argc) {
	while (optind < argc)
	    do_arg(argv[optind++]);
    } else
	do_arg(".");

    if (fixed)
	PRINTF("%d files %sfixed\n", fixed, no_op ? "would be " : "");
    else
	PRINTF("No files %sfixed\n", no_op ? "would be " : "");
    exit(SUCCESS);
    /*NOTREACHED */
}
